platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,363 @@
+Since most of the filter methods that the `Metrics` interface defines today are not used we will not introduce any additional filter methods in this proposal.
+If we see that additional methods are needed we can add them later.
+
+We will have labels that are global for a `Metrics` instance.
+The best example is the `nodeId` label.
+We do not want to add the `nodeId` label to the creation of each metric.
+Instead of that we want to add the `nodeId` label to the `Metrics` instance.
+That should happen when a `Metrics` instance is created.
+The creation of a `Metrics` instance is not public api today.
+With the implementation of this proposal a `MetricsFactory` interface should be introduced that allows to create a `Metrics` instance.
+The `MetricsFactory` interface will be defined like this:
+
+```java
+interface MetricsFactory {
+    
+    @NonNull
+    default Metrics create() {
+        return create(Set.of());
+    }
+    
+    @NonNull
+    Metrics create(@NonNull Set<Label> defaultLabels);
+}
+```
+
+The definition of the api should follow our rules regarding services as defined at https://github.com/hashgraph/hedera-services/blob/develop/platform-sdk/docs/base/service-architecture/service-architecture.md
+
+##### Examples
+
+Here are some examples how to define a metric with labels:
+
+```java
+final Label transactionTypeLabel = new Label("transactionType", "fileUpload");
+final Counter.Config config = new Counter.Config("transactionsCategory", "transactionCount").withLabels("transactionType");
+
+final Metrics metrics = ...;
+final Counter counter = metrics.getOrCreate(config);
+
+counter.withLabel(transactionTypeLabel).increment();
+```
+
+In prometheus the metric will be defined like this:
+
+```
+transactionsCategory.transactionCount{transactionType="fileUpload"}
+```
+
+#### Snapshot API
+
+The snapshot API will be extended to return the labels of a metric for a snapshot.
+By doing so any monitoring system that uses the snapshot API can use the labels.
+Today we only have 2 system that uses the snapshot API: the prometheus endpoint and the CSV exporter.
+Since no class of the snapshot API is public today (based on the internal usage of `nodeId`) we plan to move parts
+of the snapshot API to the public API.
+
+### Migration
+
+Next to extending the public API we need to migrate the current metrics to the new concept.
+
+#### Migration of the `nodeId`
+
+The `nodeId` information of a metric will be migrated to labels.
+Today we create a prometheus label out of the `nodeId` internally in our prometheus endpoint implementation.
+The same conversion can be used to create a label for the `nodeId` in future but should be moved from the prometheus endpoint to the platform code.
+The key of a metric won't change and still be defined as the combination of the `category` and `name`.
+The unique identifier of the metric is the combination key and the labels.
+That is a breaking change since today the key is the unique identifier of a metric.
```

Comment:
I'm a little confused by this. How can it be that the key and the unique identifier can be different? Will `Metrics` not support getting multiple metrics from a single category/name combination?

Reply:
if the labels are different, yes.

Assume you have a `HTTP_TIME` metric identifier. Now you can have a `method` label with the possible values of`GET` or `POST`. This will end in 2 instances.

Reply:
Edited my original comment from "not support" to "now support".

In your example of `HTTP_TIME` with a `method` label and values `GET` and `POST`, the call to `Metrics.getOrCreate(myHttpTimeMetricKey)` will return 2 metrics. How will the caller distinguish between the two and ensure the correct one gets updated?

Reply:
ahhh, got it. Here is some pseudo code:

```
MetricConfig config = new MetricConfig("category", "name").withLabels("HTTP_TYPE");

Metric metric1 = metrics.getOrCreate(config);
metric1.set(123);

final Label getLabel = new Label("HTTP_TYPE", "GET");
Metric metric2 = metrics.getOrCreate(config).withLabel(getLabel);
metric2. set(456);

final Label postLabel = new Label("HTTP_TYPE", "POST");
Metric metric3 = metrics.getOrCreate(config).withLabel(postLabel);
metric3. set(456);
```

As you can see there can be a metric with a specific label value (metric2, metric3) or even a metric without a definition for the label (metric3).

In prometheus you can than filter the 3 different metric like that:

```
category_name
category_name{http_type="GET"}
category_name{http_type="POST"}
```

How you can do more complexe queries (like get all or only specific labels) please have a look at https://prometheus.io/docs/prometheus/latest/querying/basics/#instant-vector-selectors